File: packages/gheim-py/src/gheim/detectors/calibrated.py
"""Calibrated detector — subtracts a constant `o_bias` from the O-class
logit before argmax. Originally introduced to recover the v1 model's
"single-category mode" pathology (Müller + IBAN co-occurrence flipped the
person prediction to O); the v3 / gheim-ch-560m training pipeline
(Geonames gazetteer + overlap resolver + multi-LLM consensus labelling)
largely resolved that pathology, but the calibration still buys a small
precision/recall improvement and remains the default.

Sweep on gheim-ch-560m (see eval/calibration_sweep.json,
eval/calibration_sweep_q8.json): ``o_bias=0.5`` is Pareto-clean on both
fp32 PyTorch and q8 ONNX backends.

  fp32 PyTorch (212 probe + 300-chunk test slice)
    bias 0.0  probe 91.5%  test char-F1 0.8947
    bias 0.5  probe 91.5%  test char-F1 0.8974   (+0.27pp, no probe cost)
    bias 1.0+ trades test F1 for marginal probe gains

  q8 ONNX (same suite)
    bias 0.0  probe 73.1%  test char-F1 0.8565
    bias 0.5  probe 74.5%  test char-F1 0.8608   (+1.4pp / +0.43pp)
    bias 1.0+ flattens

This is what ``default_detector()`` returns, so users get the better
behavior without having to opt in.
"""
from __future__ import annotations

import threading
from dataclasses import dataclass, field
from typing import Annotated, Any

from .base import Span, trim_span
from .local import _DEFAULT_CHUNK_TOKENS, _DEFAULT_OVERLAP_TOKENS, DEFAULT_MODEL


@dataclass
class CalibratedDetector:
    """Local model with O-logit calibration.

    Same surface as ``LocalDetector`` (lazy weight load, sliding-window
    chunking for long inputs, threadsafe), but bypasses the HF
    ``pipeline("token-classification")`` so it can intercept raw logits
    and apply ``o_bias`` before argmax. The aggregation that follows is
    a from-scratch reimplementation of ``aggregation_strategy="simple"``:
    contiguous same-label tokens collapse into one span; offsets are
    char-level via the tokenizer's ``offset_mapping``.

    Default ``o_bias=0.5`` is the Pareto-clean point from the bias sweep
    on gheim-ch-560m: small precision boost on fp32 (+0.27pp char-F1),
    larger boost on the q8 ONNX deployment backend (+0.43pp char-F1,
    +1.4pp probe perfect-rate), no observed cost on either.
    """

    model_id: Annotated[
        str, "HuggingFace model id (token-classification head)."
    ] = DEFAULT_MODEL
    o_bias: Annotated[
        float,
        "Constant subtracted from the O-class logit before argmax. 0 = "
        "uncalibrated (matches LocalDetector). Higher values recover more "
        "suppressed entities at the cost of precision. Sweet spot per "
        "eval/calibration_sweep.json: 0.5 (Pareto-clean), 1.0 (more "
        "aggressive), 1.5+ starts trading too much precision.",
    ] = 0.5
    device: Annotated[
        str, "'auto', 'cpu', 'cuda', or 'cuda:N'."
    ] = "auto"
    dtype: Annotated[Any, "Optional torch dtype override (e.g. torch.bfloat16)."] = None
    trim_whitespace: Annotated[
        bool, "Strip leading/trailing whitespace from each span."
    ] = True
    chunk_tokens: Annotated[
        int, "Sliding-window size in model tokens. Default 510 covers a "
        "512-token model with [CLS]/[SEP] reserved.",
    ] = _DEFAULT_CHUNK_TOKENS
    chunk_overlap_tokens: Annotated[
        int, "Token overlap between consecutive windows. Default 32.",
    ] = _DEFAULT_OVERLAP_TOKENS
    batch_size: Annotated[
        int, "Forward-pass batch size for the chunked + multi-input path.",
    ] = 8

    _model: Any = field(default=None, init=False, repr=False)
    _tokenizer: Any = field(default=None, init=False, repr=False)
    _id2label: dict[int, str] = field(default_factory=dict, init=False, repr=False)
    _o_id: int = field(default=0, init=False, repr=False)
    _device: str = field(default="cpu", init=False, repr=False)
    _load_lock: threading.Lock = field(default_factory=threading.Lock, init=False, repr=False)

    def _ensure_loaded(self) -> None:
        if self._model is not None:
            return
        with self._load_lock:
            if self._model is not None:
                return
            try:
                import torch
                from transformers import AutoModelForTokenClassification, AutoTokenizer
            except ImportError as e:  # pragma: no cover
                raise ImportError(
                    "CalibratedDetector requires `torch` and `transformers`. "
                    "Install with: uv add 'gheim[local]'"
                ) from e

            self._tokenizer = AutoTokenizer.from_pretrained(self.model_id, use_fast=True)
            mdl = AutoModelForTokenClassification.from_pretrained(self.model_id)
            if self.dtype is not None:
                mdl = mdl.to(dtype=self.dtype)
            resolved_device = (
                "cuda" if (self.device == "auto" and torch.cuda.is_available())
                else "cpu" if self.device == "auto"
                else self.device
            )
            mdl = mdl.to(resolved_device).eval()
            self._device = resolved_device
            self._model = mdl
            self._id2label = mdl.config.id2label
            # Resolve "O" class id (model config or fall back to label2id).
            label2id = mdl.config.label2id
            self._o_id = int(label2id.get("O", 0))

    def detect(
        self,
        text: Annotated[str, "Input text to scan."],
        model: Annotated[
            str | None, "Per-call model override; must equal self.model_id if set."
        ] = None,
    ) -> list[Span]:
        if model is not None and model != self.model_id:
            raise ValueError(
                f"CalibratedDetector was constructed with model_id={self.model_id!r}; "
                f"per-call override {model!r} not supported."
            )
        if not text:
            return []
        self._ensure_loaded()
        return self._detect_chunked(text)

    def _detect_chunked(self, text: str) -> list[Span]:
        """Chunk-aware inference. For inputs that fit in one window the
        chunking is a no-op; for longer ones we slide the window with
        overlap and reproject offsets back to the original text.
        """
        import torch  # noqa: PLC0415  — lazy import for the no-`torch` install path

        # Tokenize the whole input once with offsets so we know per-token
        # positions in the ORIGINAL text. We never feed this single tokenization
        # to the model — instead we re-encode each window separately because
        # the model expects [CLS]/[SEP] and a length ≤ chunk_tokens + 2.
        full_enc = self._tokenizer(
            text, return_offsets_mapping=True,
            add_special_tokens=False, truncation=False,
        )
        offsets_full = full_enc["offset_mapping"]
        n_tokens = len(offsets_full)

        if n_tokens <= self.chunk_tokens:
            # Single-window fast path.
            return self._detect_window(text, char_offset=0)

        # Sliding window: stride = chunk - overlap.
        stride = max(1, self.chunk_tokens - self.chunk_overlap_tokens)
        seen: set[tuple[int, int, str]] = set()
        out: list[Span] = []
        for win_start in range(0, n_tokens, stride):
            win_end = min(win_start + self.chunk_tokens, n_tokens)
            char_start = offsets_full[win_start][0]
            char_end = offsets_full[win_end - 1][1]
            sub = text[char_start:char_end]
            for sp in self._detect_window(sub, char_offset=char_start):
                key = (sp.start, sp.end, sp.label)
                if key in seen:
                    continue
                seen.add(key)
                out.append(sp)
            if win_end >= n_tokens:
                break
        out.sort(key=lambda s: (s.start, s.end))
        return out

    def _detect_window(self, text: str, char_offset: int) -> list[Span]:
        """Single forward pass on `text`, applying o_bias to the O logit
        before argmax. char_offset shifts emitted spans into the original
        text coordinate frame (used by chunked inference)."""
        import torch  # noqa: PLC0415

        enc = self._tokenizer(
            text, return_offsets_mapping=True, return_tensors="pt",
            truncation=True, max_length=self.chunk_tokens + 2,
        )
        offsets = enc.pop("offset_mapping")[0].tolist()
        enc = {k: v.to(self._device) for k, v in enc.items()}
        with torch.no_grad():
            logits = self._model(**enc).logits[0].cpu()
        # The calibration knob: subtract o_bias from O class logit.
        if self.o_bias != 0.0:
            logits[:, self._o_id] -= float(self.o_bias)
        pred_ids = logits.argmax(dim=-1).tolist()

        # BIO aggregation: collapse contiguous same-category tokens into
        # one span. This mirrors transformers' "simple" aggregation but
        # without the post-softmax pruning that the pipeline does.
        spans: list[Span] = []
        cur_start, cur_end, cur_cat = None, None, None
        for (s, e), pid in zip(offsets, pred_ids):
            if s == e:  # special token (CLS, SEP, pad)
                continue
            lab = self._id2label[int(pid)]
            cat = None if lab == "O" else lab.split("-", 1)[1]
            if cat == cur_cat and cat is not None:
                cur_end = e
            else:
                if cur_cat is not None:
                    spans.append(self._make_span(text, cur_start, cur_end, cur_cat,
                                                  char_offset))
                cur_start, cur_end, cur_cat = s, e, cat
        if cur_cat is not None:
            spans.append(self._make_span(text, cur_start, cur_end, cur_cat,
                                          char_offset))
        return spans

    def _make_span(
        self, text: str, start: int, end: int, label: str, char_offset: int,
    ) -> Span:
        if self.trim_whitespace:
            start, end = trim_span(text, start, end)
        return Span(
            start=start + char_offset,
            end=end + char_offset,
            label=label,
            text=text[start:end],
            score=1.0,  # we collapsed across the softmax; no per-span score
        )

File: packages/gheim-js/src/detectors/calibrated.ts
/**
 * Calibrated detector — subtracts a constant `oBias` from the O-class
 * logit before argmax. Originally introduced to recover the v1 model's
 * "single-category mode" pathology (Müller + IBAN co-occurrence flipped
 * the person prediction to O); the v3 / gheim-ch-560m training pipeline
 * largely resolved that pathology, but the calibration still buys a
 * small precision/recall improvement and remains the default.
 *
 * Sweep on gheim-ch-560m (see eval/calibration_sweep.json):
 * `oBias=0.5` is the Pareto-clean default — probe perfect-rate is flat
 * at 91.5% across bias 0.0 and 0.5, and test char-F1 lifts by 0.27pp.
 * Higher biases (1.0–1.5) trade test F1 for marginal probe gains.
 *
 * A separate sweep on the int8 ONNX path via `transformers.js`
 * (eval/calibration_sweep_q8.json) showed a much larger calibration
 * effect (+1.4pp probe, +0.43pp char-F1 at 0.5). That gap is not a
 * property of the int8 quantisation — Python `onnxruntime` on the same
 * int8 file gives 91.0% probe perfect-rate, essentially fp32-equivalent
 * — but of a runtime divergence inside `onnxruntime-web` / WASM that
 * the calibration partially papers over. The reliable fix is to use the
 * `model_fp16.onnx` variant when WebGPU is available (which is what
 * LocalDetector's default `dtype: "auto"` does); int8 should be treated
 * as a low-RAM fallback. See eval/q8_quality_report.md for the
 * diagnostic.
 *
 * `defaultDetector()` returns this with oBias=0.5, so users get the
 * better behavior automatically.
 */
import type { Detector, Span } from "./base.ts";
import type { LocalDetectorLoadEvent } from "./local.ts";

const DEFAULT_MODEL = "joelbarmettler/gheim-ch-560m";
const DEFAULT_CHUNK_TOKENS = 510;
const DEFAULT_OVERLAP_TOKENS = 32;

/** Constructor options for {@link CalibratedDetector}. */
export interface CalibratedDetectorOptions {
  /** HuggingFace model id (must be a token-classification head). */
  model?: string;
  /**
   * Constant subtracted from the O-class logit before argmax. Default
   * `0.5` is the Pareto-clean point on the reference (fp32/fp16) path:
   *
   *   - oBias=0.0  uncalibrated (matches the raw model)
   *   - oBias=0.5  +0.27pp test char-F1, no probe cost (91.5% perfect)
   *   - oBias=1.0+ marginal probe gains, test F1 starts to fall off
   *
   * See eval/calibration_sweep.json.
   */
  oBias?: number;
  /** Backend selector. See {@link LocalDetectorOptions.device}. */
  device?: string;
  /** Quantization dtype: "fp32" | "fp16" | "q8" | "q4" | "auto". */
  dtype?: string;
  /** Sliding-window size in model tokens (default 510). */
  chunkTokens?: number;
  /** Overlap between consecutive windows (default 32). */
  chunkOverlapTokens?: number;
  /** Strip leading/trailing whitespace from each emitted span. Default true. */
  trimWhitespace?: boolean;
  /** Optional progress callback during model load (mirrors LocalDetector). */
  onProgress?: (event: LocalDetectorLoadEvent) => void;
}

interface TransformersJsModule {
  AutoTokenizer: { from_pretrained: (id: string, opts?: unknown) => Promise<unknown> };
  AutoModelForTokenClassification: {
    from_pretrained: (id: string, opts?: unknown) => Promise<unknown>;
  };
}

interface TokenizerCallable {
  (text: string, opts: {
    return_tensors?: string;
    truncation?: boolean;
    max_length?: number;
    return_offsets_mapping?: boolean;
    add_special_tokens?: boolean;
  }): Promise<TokenizerOutput> | TokenizerOutput;
}

interface TokenizerOutput {
  input_ids?: Tensor;
  attention_mask?: Tensor;
  // transformers.js v4 returns plain objects with these fields. Note:
  // offset_mapping is NOT supported (TODO in transformers.js source).
}

interface Tensor {
  data: ArrayLike<number> | Int32Array | Float32Array | BigInt64Array;
  dims: number[];
  tolist?: () => unknown;
}

interface ModelCallable {
  (inputs: Record<string, Tensor>): Promise<{ logits: Tensor }>;
  config: { id2label: Record<string, string>; label2id: Record<string, number> };
}

/** Trim leading/trailing whitespace from a [start, end) slice of `text`. */
function trimSpan(text: string, start: number, end: number): [number, number] {
  while (start < end && /\s/.test(text[start] ?? "")) start++;
  while (end > start && /\s/.test(text[end - 1] ?? "")) end--;
  return [start, end];
}

/** CalibratedDetector — local model + O-logit bias. */
export class CalibratedDetector implements Detector {
  readonly model: string;
  readonly oBias: number;
  readonly device: string;
  readonly dtype: string;
  readonly chunkTokens: number;
  readonly chunkOverlapTokens: number;
  readonly trimWhitespace: boolean;
  /** Backend actually used after `load()` resolves; `null` before. */
  actualDevice: string | null = null;
  /** Dtype actually used after `load()` resolves; `null` before. Differs
   *  from {@link dtype} only when `dtype: "auto"` was requested. */
  actualDtype: string | null = null;
  private _model: ModelCallable | null = null;
  private _tokenizer: TokenizerCallable | null = null;
  private _id2label: Record<number, string> = {};
  private _oId = 0;
  private _loadPromise: Promise<void> | null = null;
  private readonly _onProgress?: (event: LocalDetectorLoadEvent) => void;

  constructor(opts: CalibratedDetectorOptions = {}) {
    this.model = opts.model ?? DEFAULT_MODEL;
    this.oBias = opts.oBias ?? 0.5;
    this.device = opts.device ?? "auto";
    this.dtype = opts.dtype ?? "auto";
    this.chunkTokens = opts.chunkTokens ?? DEFAULT_CHUNK_TOKENS;
    this.chunkOverlapTokens = opts.chunkOverlapTokens ?? DEFAULT_OVERLAP_TOKENS;
    this.trimWhitespace = opts.trimWhitespace ?? true;
    this._onProgress = opts.onProgress;
    if (this.chunkOverlapTokens >= this.chunkTokens) {
      throw new Error(
        `chunkOverlapTokens (${this.chunkOverlapTokens}) must be smaller than ` +
          `chunkTokens (${this.chunkTokens})`,
      );
    }
  }

  async load(): Promise<void> {
    if (this._model) return;
    if (this._loadPromise) return this._loadPromise;
    this._loadPromise = (async () => {
      const mod = await loadTransformersJs();
      // Pick a backend (auto-fallback identical to LocalDetector). Node
      // transformers.js uses onnxruntime-node which doesn't speak "wasm" —
      // fall back to "cpu" instead. We discriminate via `typeof window`
      // (Node 21+ ships a global `navigator`, so that no longer works).
      let preferredDevice: string = this.device;
      if (this.device === "auto") {
        const isBrowser =
          typeof (globalThis as { window?: unknown }).window !== "undefined";
        const nav = typeof navigator !== "undefined"
          ? (navigator as { gpu?: { requestAdapter: () => Promise<unknown> } })
          : null;
        preferredDevice = isBrowser ? "wasm" : "cpu";
        if (nav?.gpu?.requestAdapter) {
          try {
            const adapter = await nav.gpu.requestAdapter();
            if (adapter) preferredDevice = "webgpu";
          } catch {
            /* keep the platform fallback */
          }
        }
      }
      // Resolve `dtype: "auto"` to a concrete file: fp16 on WebGPU (byte-
      // equivalent to fp32 on the forensic probe), q8 elsewhere (small,
      // safe fallback, but degrades on Swiss edge cases via the
      // onnxruntime-web WASM kernel — see eval/q8_quality_report.md).
      const resolvedDtype = this.dtype === "auto"
        ? (preferredDevice === "webgpu" ? "fp16" : "q8")
        : this.dtype;
      const tokenizer = (await mod.AutoTokenizer.from_pretrained(this.model)) as TokenizerCallable;
      const model = (await mod.AutoModelForTokenClassification.from_pretrained(this.model, {
        device: preferredDevice,
        dtype: resolvedDtype,
        progress_callback: this._onProgress
          ? (raw: RawProgress) => this._emitProgress(raw, preferredDevice)
          : undefined,
      })) as ModelCallable;
      this._tokenizer = tokenizer;
      this._model = model;
      this.actualDtype = resolvedDtype;
      this._id2label = model.config.id2label as unknown as Record<number, string>;
      const label2id = model.config.label2id;
      this._oId = label2id?.["O"] ?? 0;
      this.actualDevice = preferredDevice;
    })();
    return this._loadPromise;
  }

  async detect(text: string, opts?: { model?: string }): Promise<Span[]> {
    if (opts?.model && opts.model !== this.model) {
      throw new Error(
        `CalibratedDetector was constructed with model='${this.model}'; ` +
          `per-call model override ('${opts.model}') not supported.`,
      );
    }
    if (!text) return [];
    await this.load();
    return this._detectChunked(text);
  }

  private async _detectChunked(text: string): Promise<Span[]> {
    // transformers.js's tokenizer doesn't return offset_mapping, so we
    // can't slice token-windows precisely. Fall back to char-window
    // chunking with overlap (mirrors LocalDetector's character path).
    if (text.length <= 1500) {
      return await this._detectWindow(text, 0);
    }
    const stride = 1500 - 200; // ~chunkTokens × ~3 chars/tok with overlap
    const seen = new Set<string>();
    const out: Span[] = [];
    for (let s = 0; s < text.length; s += stride) {
      const e = Math.min(s + 1500, text.length);
      const sub = text.slice(s, e);
      const subSpans = await this._detectWindow(sub, s);
      for (const sp of subSpans) {
        const key = `${sp.start}|${sp.end}|${sp.label}`;
        if (seen.has(key)) continue;
        seen.add(key);
        out.push(sp);
      }
      if (e >= text.length) break;
    }
    out.sort((a, b) => a.start - b.start || a.end - b.end);
    return out;
  }

  private async _detectWindow(text: string, charOffset: number): Promise<Span[]> {
    const tokenizer = this._tokenizer! as unknown as {
      (texts: string[] | string, opts: unknown): Promise<TokenizerOutput>;
      decode: (ids: number[], opts: { skip_special_tokens?: boolean }) => string;
    };
    const model = this._model!;
    const enc = await tokenizer([text], {
      padding: true,
      truncation: true,
      max_length: this.chunkTokens + 2,
    });
    const inputs: Record<string, Tensor> = {};
    if (enc.input_ids) inputs["input_ids"] = enc.input_ids;
    if (enc.attention_mask) inputs["attention_mask"] = enc.attention_mask;
    const out = await model(inputs);
    const logits = readLogits(out.logits); // [seq, num_labels]
    // Apply the O-bias before argmax.
    if (this.oBias !== 0) {
      for (let t = 0; t < logits.length; t++) {
        const row = logits[t];
        if (row === undefined) continue;
        const cur = row[this._oId];
        if (cur !== undefined) {
          row[this._oId] = cur - this.oBias;
        }
      }
    }
    // Argmax per token; build (token_id, label) list.
    const inputIds = readInputIds(enc.input_ids);
    const tokenLabels: Array<{ id: number; label: string }> = [];
    for (let t = 0; t < logits.length; t++) {
      const row = logits[t];
      const id = inputIds[t];
      if (!row || id === undefined) continue;
      let maxIdx = 0;
      let maxVal = row[0]!;
      for (let i = 1; i < row.length; i++) {
        if (row[i]! > maxVal) {
          maxVal = row[i]!;
          maxIdx = i;
        }
      }
      tokenLabels.push({ id, label: this._id2label[maxIdx] ?? "O" });
    }
    // BIO aggregation: collapse consecutive same-category tokens. For each
    // collapsed group, decode its tokens to a surface form, then locate
    // that surface in the original `text` via cursor-tracked substring
    // search (transformers.js doesn't expose char offsets — same trick
    // LocalDetector uses for its substring recovery).
    const spans: Span[] = [];
    let cursor = 0;
    let groupIds: number[] = [];
    let groupCat: string | null = null;
    const flush = () => {
      if (groupCat === null || groupIds.length === 0) return;
      const surface = normalizeDecoded(
        tokenizer.decode(groupIds, { skip_special_tokens: true }),
      );
      groupIds = [];
      const cat = groupCat;
      groupCat = null;
      if (!surface) return;
      const idx = text.indexOf(surface, cursor);
      if (idx === -1) return;
      let s = idx;
      let e = idx + surface.length;
      cursor = e;
      if (this.trimWhitespace) [s, e] = trimSpan(text, s, e);
      if (s >= e) return;
      spans.push({
        start: s + charOffset,
        end: e + charOffset,
        label: cat,
        text: text.slice(s, e),
        score: 1.0,
      });
    };
    for (const { id, label } of tokenLabels) {
      const cat = label === "O" ? null : label.split("-", 2)[1] ?? null;
      if (cat === groupCat && cat !== null) {
        groupIds.push(id);
      } else {
        flush();
        if (cat !== null) {
          groupIds.push(id);
          groupCat = cat;
        }
      }
    }
    flush();
    return spans;
  }

  private _emitProgress(raw: RawProgress, device: string): void {
    if (!this._onProgress) return;
    const event: LocalDetectorLoadEvent = {
      phase: (raw.status as LocalDetectorLoadEvent["phase"]) ?? "progress",
      device,
    };
    if (raw.file) event.file = raw.file;
    if (raw.status === "progress") {
      if (raw.loaded != null) event.loadedBytes = raw.loaded;
      if (raw.total != null) event.totalBytes = raw.total;
      if (raw.progress != null) {
        const norm = raw.progress > 1 ? raw.progress / 100 : raw.progress;
        event.fraction = Math.max(0, Math.min(1, norm));
      }
    }
    this._onProgress(event);
  }
}

interface RawProgress {
  status: string;
  file?: string;
  progress?: number;
  loaded?: number;
  total?: number;
}

async function loadTransformersJs(): Promise<TransformersJsModule> {
  try {
    return (await import("@huggingface/transformers")) as unknown as TransformersJsModule;
  } catch {
    throw new Error(
      "CalibratedDetector requires `@huggingface/transformers`. " +
        "Install with: bun add @huggingface/transformers",
    );
  }
}

/** Read a flat input_ids array from a [batch=1, seq] tensor. */
function readInputIds(tensor: Tensor | undefined): number[] {
  if (!tensor) return [];
  if (typeof tensor.tolist === "function") {
    const list = tensor.tolist();
    if (Array.isArray(list) && Array.isArray(list[0])) {
      return (list[0] as unknown[]).map((x) => Number(x));
    }
    if (Array.isArray(list)) return (list as unknown[]).map((x) => Number(x));
  }
  const dims = tensor.dims;
  const seq = dims[dims.length - 1] ?? 0;
  const data = tensor.data;
  const out: number[] = [];
  for (let i = 0; i < seq; i++) out.push(Number(data[i]));
  return out;
}

/** Strip BPE wordpiece markers and outer whitespace from a decoded surface. */
function normalizeDecoded(s: string): string {
  return s.replace(/##/g, "").trim();
}

/** Read a [seq, num_labels] logits tensor as a number[][]. */
function readLogits(tensor: Tensor): number[][] {
  if (typeof tensor.tolist === "function") {
    const list = tensor.tolist();
    if (Array.isArray(list) && Array.isArray(list[0]) && Array.isArray((list[0] as unknown[])[0])) {
      // [batch, seq, num_labels] — take first batch.
      return (list as unknown as number[][][])[0]!;
    }
    if (Array.isArray(list) && Array.isArray(list[0])) {
      return list as number[][];
    }
  }
  // Fallback: read raw data + dims.
  const dims = tensor.dims;
  const seq = dims[dims.length - 2] ?? 0;
  const cls = dims[dims.length - 1] ?? 0;
  const data = tensor.data;
  const out: number[][] = [];
  for (let t = 0; t < seq; t++) {
    const row: number[] = [];
    for (let c = 0; c < cls; c++) {
      row.push(Number(data[t * cls + c]));
    }
    out.push(row);
  }
  return out;
}
